File: golang/graph_dfs/keys_and_rooms.go
package main

func canVisitAllRooms(rooms [][]int) bool {
	visited := make([]bool, len(rooms))
	visited[0] = true

	queue := []int{0}

	for len(queue) > 0 {
		room := queue[0]
		queue = queue[1:]

		for _, key := range rooms[room] {
			if !visited[key] {
				visited[key] = true

				queue = append(queue, key)
			}
		}
	}

	for _, v := range visited {
		if !v {
			return false
		}
	}

	return true
}

File: java/src/main/java/com/sanedge/interview/graph_dfs/KeysAndRooms.java
package com.sanedge.interview.graph_dfs;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class KeysAndRooms {

    public boolean canVisitAllRoom(List<List<Integer>> rooms) {
        int n = rooms.size();

        boolean[] visited = new boolean[n];
        visited[0] = true;

        Queue<Integer> queue = new LinkedList<>();
        queue.offer(0);

        while (!queue.isEmpty()) {
            int room = queue.poll();

            for (int key : rooms.get(room)) {
                if (!visited[key]) {
                    visited[key] = true;
                    queue.offer(key);
                }
            }
        }

        

        for (boolean v : visited) {
            if (!v) {
                return false;
            }
        }

        return true;
    }
}
